tests/mocks/Unity/Application/SurfaceManager.cpp

```diff
@@ -35,6 +35,9 @@ SurfaceManager *SurfaceManager::m_instance = nullptr;
 
 SurfaceManager *SurfaceManager::instance()
 {
+    if (m_instance == nullptr) {
+        auto inst = new SurfaceManager();
+    }
```

Comment:
This causes a warning as the compiler thinks it is a useless statement. Obviously it isn't, as it ends up setting m_instance as a side effect.

In more interesting news, somewhere around here is the cause of the failures in the tst_ShellWithPin tests. When built with debugging enabled, the Q_ASSERT below (`Q_ASSERT(m_instance == nullptr)`) fails:

```
QFATAL : qmltestrunner::ShellWithPin::test_bfbOnLockedDevice() ASSERT: "m_instance == nullptr" in file /home/ubuntu/unity8/tests/mocks/Unity/Application/SurfaceManager.cpp, line 45
FAIL!  : qmltestrunner::ShellWithPin::test_bfbOnLockedDevice() Received a fatal error.
```

It appears that we may have two instances of a SurfaceManager in flight during the tests... I'm not sure why that would be, though. Maybe it's okay to have more than one instance of a SurfaceManager and the instance shouldn't be static. Maybe we need to do something to ensure that the old one is destroyed.

Reply:
Wait, resync. The test failure when *not* built with debugging is the new application never gets a Surface, leading to the wrong thing being focused. So having two SurfaceManagers definitely wreaks havoc.

Reply:
There must be the code somewhere that's supposed to create an instance of this class, be it C++ or QML. In what test does this cause a failure?

Reply:
No, there is only ever one SurfaceManager instance at a time in the tests. It's a static variable, which is why the assertion to check there is only one of them exists.

I've investigated what it would take to make SurfaceManager::instance() a non-static object, which would be best, but unfortunately a lot of other mocks depend on its stasis. We'd need to implement much more of the Mir stack the same way it's implemented in production to fix that... And once that's done, there are other mocks implemented in this static way.

So, we need to ensure that there is only ever one SurfaceManager in flight. Why isn't there one now?

The answer is... race conditions. QML object destruction is unordered and asynchronous, so in the tests (specifically in ShellWithPin), it's possible to destroy one Shell instance but create a new one before all of the old one's child objects are removed. If the SurfaceManager is the object that happens to remain in flight... boom.

Right now I'm trying to focus on a way to ensure the old object tree is deleted before the new one is instantiated. Turns out that's really difficult... If you call `.destroy()` on an object, the deletion is queued for whenever the event loop is ready to do it. Even running `gc()` manually doesn't help for some unholy reason.

So... now I'm working on using [TestCase.createTemporaryObject](https://doc.qt.io/qt-5/qml-qttest-testcase.html#createTemporaryQmlObject-method), which seems to be the intended way to solve this problem.

Reply:
I've pulled in the changes from @UniversalSuperBox related to this, so I suppose this can be resolved now?

Reply:
Should this not be removed then?